### PR TITLE
COUNT queries can be very slow

### PR DESCRIPTION
**Pull request by diogogithub**
As in https://notabug.org/diogo/gnu-social/commit/586fb5a5175d7a10f5f78dd026434e48202e5451

Some context from #social@irc.freenode.org:

```
<XRevan86> > I've added chimo's Nodeinfo plugin as a default plugin on GS
<XRevan86> diogo: Heavy queries is why ldavg doesn't have it.
<dansup> who runs loadaverage
<XRevan86> dansup: I do.
<XRevan86> dansup: Unless you mean hosting, that's being done by maiyannah, and the domain is owned by pztrn
<aab> up201705417, dansup: good idea. I've had Nodeinfo enabled for a long time... don't know what "heavy queries" mean in this context... Too many connections?
<TakinOver> aab: I think heavy queries probably refers to its database usage. LoadAverage isn't a 2-3 active users instance.
<aab> ok, but... with chimo's plugin you get to show general info about your node, not about every single user, right?
<aab> i mean, look at "https://khp.ignorelist.com/main/nodeinfo/2.0". Isn't the same if i had 3000 users?
<TakinOver> I haven't tried chimo's plugin yet. Is it anything like pztrn's old gstools plugin?
<aab> https://github.com/chimo/gs-nodeinfo
<aab> i don't know about pztrn, sorry...
<aab> i installed because it seems to be a "standarized" way to identifiy your node to other nodes on federation
<aab> and appear on https://fediverse.network/, to show GS is not dead XD
<TakinOver> fediverse.network seems to try lots of different ways to identify and summarize. When they first appeared in the logs, I almost blocked them because there were so many different requests and they happened so frequently.
<diogo> TakinOver: yup, they are a bit abusive indeed...
<aab> ah, ok. I saw a request per day, maybe they changed it by the time i installed the plugin...
<aab> i guess you meant this https://github.com/andstatus/andstatus/issues/84
<TakinOver> aab: Yes, the gstools plugin worked with gstools.org, but both are currently abandoned.
<aab> ok
<XRevan86> I suppose I should give https://code.chromic.org/chimo/gs-nodeinfo another chance, that way I'll have something more concrete to show.
<XRevan86> diogo: Enabled it. Brace yourselves
<aab> so... it depends on the number of accounts? Or what else?
<XRevan86> > SELECT count(notice.uri) as DATAOBJECT_NUM FROM notice WHERE ( reply_to IS NULL ) AND ( notice.is_local = 1 ):
<aab> ah, ok, number of notices
<XRevan86> aab: Basically, ldavg has too many notices for its own good :)
<diogo> it does run... 5 queries in the metadata section
<diogo> let me see how to fix that, one sec...
<XRevan86> even though it counts only local ones, it has to comb through all of them
<xmpp-gnu> ***XRevan86 terminates the plugin
<aab> sorry if i miss something, i'm trying to understand: so, you are saying, every query of nodeinfo counts all notices on the node?
<aab> every time?
<XRevan86> aab: It looks that way.
<aab> well, i have no programming skills, so maybe this is dumb, but... Shouldn't database be aware of the number of notices it has in every moment, so it doesn't have to recount on every query?
<xmpp-gnu> [colegota] https://bin.disroot.org/?d7643214e9918ce5#vTkKuyfrmmL3D2p2f7c7cHN4oXfHc5FKocMldC6m16w=
<xmpp-gnu> [colegota] A collection of log entries with "gnusocial"
<diogo> aab: it actually depends in the database engine, some do store, others don't, the one GS is using (innoDB) doesn't
<diogo> aab: https://www.percona.com/blog/2006/12/01/count-for-innodb-tables/
<diogo> also note that there are some constraints like "notice has to be local"
<diogo> so, it really has to go through all of them
<diogo> in order to maintain statistics in a performant way we have to create counters in GS
<aab> ok... so i haven't noticed a great impact because of low number of notices... thanks!
<XRevan86> https://code.chromic.org/chimo/gs-nodeinfo/src/commit/6ee5c3eb6d2aad0aa321d6aff534e4ed5a5a50ed/actions/nodeinfo_2_0.php#L66
<XRevan86> Actually… it counts it twice :)
<XRevan86> first notices, then replies
<XRevan86> in two separate queries
<aab> 😲
<XRevan86> diogo: I wonder if it's possible to make $notices->count(); count on id's instead of uri's in the plugin. Then notice_created_id_is_local_idx applies and such a query only takes 5 minutes instead of… a lot, it hasn't finished yet
<diogo> 5 minutes already is a lot
<diogo> I was creating a table to keep the counts...
<XRevan86> it is still going!
<XRevan86> I think it's already half an hour
<diogo> that's a really huge database
<XRevan86> diogo: 9 million notices in total
<XRevan86> Maybe if I create an index just for nodeinfo…
<XRevan86> CREATE INDEX notice_uri_replyto_is_local_idx ON notice (uri, reply_to, is_local); is going to take a bit of time :)
<diogo> it will probably help, but I still think it would be better to have this answer in O(1)
<XRevan86> diogo: Well, I'm just curious how much it'll take.
<diogo> you need the value anyway so, go for it :)
<XRevan86> hm, it affects the website…
<XRevan86> ugh, I guess I'll wait
<XRevan86> though whatever
```

### CODE REVIEW

**Comment by diogogithub** [comment]
How pixelfed is doing: https://github.com/pixelfed/pixelfed/blob/dev/app/Http/Controllers/FederationController.php#L91-L159

File: actions/nodeinfo_2_0.php [hunk 1]
```diff
@@ -36,77 +36,120 @@ function getActivePluginList()
...
                 $protocols[] = 'ostatus';
             }
-
             if ($xmppEnabled) {
                 $protocols[] = 'xmpp';
             }
         }
-        Event::handle('EndNodeInfoProtocols', array(&$protocols));
-
+        Event::handle('EndNodeInfoProtocols', array(&$protcols));
```
Comment (chimo) [inline actions/nodeinfo_2_0.php]: Looks like the previously fixed typo is back with this one

Comment (diogogithub) [reply actions/nodeinfo_2_0.php]: ooops 😓, just fixed!

**Comment by chimo** [comment]
Thanks, this is great

**Pull request merged**